### ISSUE TITLE
Update contrib documentation to prefer "squash & merge"

### ISSUE THREAD

**Issue by arostamianfar** [DataBiosphere/toil#2816]
**tl;dr: prefer using "Squash and merge" when merging PRs into master whenever possible.**

As discussed offline with @adamnovak, we're proposing to use Github's "Squash and merge" option when merging PRs into master rather than "Rebase & merge" for *most* PRs.

Pros:
* The most important benefit is that it makes the commit history on master much more readable as a single commit is created at the time of merge rather than commits being scattered throughout time. As a result, it makes debugging in case of a breakage much easier (i.e. when listing all main changes from last release that may have caused an issue).
* It still preserves detailed history in the PR itself (e.g. how different comments were addressed) while not polluting master's commit history with "fix typo" style of commits.
* Reverting the PR can be done by reverting a single commit.

Cons:
* Granular history of commits within the PR are lost on master. In most cases, these are fixes to the original commit, which don't add much value anyway. However, it can become problematic if a PR contains unrelated commits (we should try to avoid these whenever possible).
* When squashing commits from multiple authors, one needs to remember to add "Co-authors" using https://github.blog/2018-01-29-commit-together-with-co-authors/ .
* ~~[Looks like](https://github.com/isaacs/github/issues/1368) PR authorship is granted to the person who clicks the "Squash & merge" button rather than the person who actually worked on the commit, which is definitely not desirable for 3rd party contributions. GitHub has a workaround using "co-authors" for this: https://github.blog/2018-01-29-commit-together-with-co-authors/~~

**Proposed decision**:
* Prefer "Squash and merge" when the PR contains a single "unit of work" (i.e. if one were to rewrite the PR from scratch with all the fixes included, they would have one commit for the entire PR).
* When merging commits from multiple authors, include `Co-authored-by:` in the commit message or continue to use "Rebase & merge" if the PR contains a small number of commits.

┆Issue is synchronized with this [Jira Task](https://ucsc-cgl.atlassian.net/browse/TOIL-434)
┆Issue Number: TOIL-434

**Comment by adamnovak** [comment]
@arostamianfar I looked at the issue thread you mentioned to support the use of `Co-authored-by:`, and it looks to me like it is complaining that Github is rewriting the original author's name and e-mail to *whatever is associated with their GitHub account*, not to the name and e-mail of the person clicking merge. Still annoying, but not to the point where it's attributing commits to the wrong people.

Is that your interpretation too?

**Comment by arostamianfar** [comment]
@adamnovak Good point! I think you're right. It looks like the `Co-authored by:` is only required if multiple authors worked on the PR at various times (based on https://github.com/isaacs/github/issues/1303#issuecomment-486089549 ). If only one author contributed to the PR, then the squashed commit would be associated with the actual author of the commit and not the person who ended up merging it. I updated the comment above to reflect this.
Will send a PR to update the docs soon.
